### DISCUSSION TITLE
Serializer overloads for Struct duals of Tables

### DISCUSSION THREAD

**Discussion by Astn** [jamescourtney/FlatSharp#211] [Ideas]
# The Problem

Before I can serialize, I must create the objects to be serialized. In many cases these objects must be heap allocated classes. So this is an attempt at improving use cases like parsing a file, or receiving data in another format before getting it into a buffer.

## A Solution

Generate struct duals of Tables that can be used in the serialization process.

### Some data

- Class1000 creates 1000 objects (heap) representing rows with around 100 cells (class) each.
- Struct1000 creates 1000 objects (heap) representing rows with around 100 cells (struct) each.
- DataProvider1000 see DataProvider discussion #213 
```
|           Method |      Mean |     Error |    StdDev |    StdErr |       Min |       Max |        Q1 |    Median |        Q3 |   Op/s |      Gen 0 |     Gen 1 | Gen 2 |    Allocated |
|----------------- |----------:|----------:|----------:|----------:|----------:|----------:|----------:|----------:|----------:|-------:|-----------:|----------:|------:|-------------:|
|        Class1000 | 29.386 ms | 0.5695 ms | 0.7796 ms | 0.1529 ms | 28.495 ms | 31.349 ms | 28.699 ms | 29.122 ms | 30.024 ms |  34.03 | 24031.2500 | 3937.5000 |     - | 147251.43 KB |
|       Struct1000 |  3.048 ms | 0.0607 ms | 0.1079 ms | 0.0171 ms |  2.917 ms |  3.310 ms |  2.944 ms |  3.025 ms |  3.127 ms | 328.13 |     3.9063 |         - |     - |     25.15 KB |
| DataProvider1000 |  1.408 ms | 0.0259 ms | 0.0419 ms | 0.0072 ms |  1.360 ms |  1.508 ms |  1.382 ms |  1.394 ms |  1.410 ms | 710.17 |    17.5781 |         - |     - |    118.77 KB |

```

Sample Benchmark
```C#

[Benchmark]
        public void Class1000()
        {
            var kvs = new List<KeyValue>();
            var map = new Map
            {
                Items = kvs
            };
            var buffer = new byte[1024];
            var now = DateTime.Now.ToFileTimeUtc();
            var rowsCount = rows.Length;
            var headerCount = header.Length;
            for (int r = 0; r < 1000; r++)
            {
                var row = rows[r % rowsCount];
                for (int col = 0; col < row.Length && col < headerCount; col++)
                {
                    KeyValue kv = new KeyValue { Key = header[col], Value = row[col], Timestamp = now };
                    if (!string.IsNullOrWhiteSpace(kv.Value.String))
                    {
                        kvs.Add(kv);
                    }
                }
                // serialize Kvs
                //var writtenSize = serialize(map, buffer);
                
                kvs.Clear();
            }
        }
        
        [Benchmark]
        public void Struct1000()
        {
            var kvs = new List<(string,string,long)>();
            var buffer = new byte[1024];
            var now = DateTime.Now.ToFileTimeUtc();
            var rowsCount = rows.Length;
            var headerCount = header.Length;
            for (int r = 0; r < 1000; r++)
            {
                var row = rows[r % rowsCount];
                for (int col = 0; col < row.Length && col < headerCount; col++)
                {
                    var kv = (Key : header[col], Value : row[col], Timestamp : now );
                    if (!string.IsNullOrWhiteSpace(kv.Value))
                    {
                        kvs.Add(kv);
                    }
                }
                // serialize Kvs
                //var writtenSize = serialize(map, buffer);
                
                kvs.Clear();
            }
        }
```

**Comment by jamescourtney** [comment]
This is a fair point, and something that I'm guilty of overlooking. I haven't spent much time thinking about how to get to the point of giving data to the library.

I don't think FlatSharp is unusual in that regard (most serializers I'm familiar with require you to provide a fully populated object graph). An interesting exception is the Google FlatBuffer library, which exposes a builder pattern that deals with problems such as this.

To be contrarian for a moment, have you considered pooling those objects, rather than making new ones each time? FlatSharp won't use them any more after the serialize operation finishes, so you could do something along the lines of:

```c#
       private static readonly ThreadLocal<Queue<KeyValue>> SimplePool = new(() => new Queue<KeyValue>());

        public void Class1000()
        {
            var kvs = new List<KeyValue>();
            var map = new Map
            {
                Items = kvs
            };
            var buffer = new byte[1024];
            var now = DateTime.Now.ToFileTimeUtc();
            var rowsCount = rows.Length;
            var headerCount = header.Length;
            var pool = SimplePool.Value;
            for (int r = 0; r < 1000; r++)
            {
                var row = rows[r % rowsCount];
                for (int col = 0; col < row.Length && col < headerCount; col++)
                {
                    if (string.IsNullOrWhiteSpace(row[col].String)) continue;

                    if (!pool.TryDequeue(out KeyValue kv))
                    {
                         kv = new KeyValue();
                    }

                    kv.Key = header[col];
                    kv.Value = row[col];
                    kv.Timestamp = row;
                    kvs.Add(kv);
                }
                // serialize Kvs
                //var writtenSize = serialize(map, buffer);

                foreach (var item in kvs)
                {
                     pool.Enqueue(item);
                }         
                
                kvs.Clear();
            }
        }
```

I'm sure there are even more clever ways you could do the pooling. The benchmark is missing some code so I can't run it on my box.

**Reply by Astn** [reply]
> I haven't spent much time thinking about how to get to the point of giving data to the library.

That's fine, your library is already better then Google's. 
Even though Google trys to provide something to address this, their approach turned out awful. So if we do something to address this, we should learn from them and not do what they did. 🤷

**Reply by Astn** [reply]
> have you considered pooling those objects, rather than making new ones each time?

Strangely, my pooled version performs worse than non pooled. But I did include that in the benchmark branch I linked below.

**Comment by Astn** [comment]
@jamescourtney  I've create a branch with these benchmarks in it at #214 so you can play around if you like.

```
|            Method |      Mean |     Error |    StdDev |    StdErr |       Min |       Max |        Q1 |    Median |        Q3 |   Op/s | Ratio | RatioSD |     Gen 0 |    Gen 1 | Gen 2 |   Allocated |
|------------------ |----------:|----------:|----------:|----------:|----------:|----------:|----------:|----------:|----------:|-------:|------:|--------:|----------:|---------:|------:|------------:|
|         Class1000 | 19.811 ms | 0.3868 ms | 0.6568 ms | 0.1080 ms | 18.493 ms | 21.154 ms | 19.369 ms | 19.748 ms | 20.322 ms |  50.48 | 13.65 |    0.61 | 7656.2500 | 468.7500 |     - | 47001.43 KB |
| Class1000Recycled | 31.993 ms | 0.6395 ms | 1.1200 ms | 0.1793 ms | 30.108 ms | 34.840 ms | 31.111 ms | 32.048 ms | 32.622 ms |  31.26 | 22.09 |    1.09 | 4656.2500 | 656.2500 |     - | 28611.94 KB |
|        Struct1000 |  3.085 ms | 0.0615 ms | 0.1078 ms | 0.0173 ms |  2.922 ms |  3.320 ms |  2.964 ms |  3.131 ms |  3.169 ms | 324.13 |  2.10 |    0.07 |    3.9063 |        - |     - |    25.15 KB |
|  DataProvider1000 |  1.458 ms | 0.0282 ms | 0.0404 ms | 0.0076 ms |  1.381 ms |  1.506 ms |  1.428 ms |  1.476 ms |  1.487 ms | 686.02 |  1.00 |    0.00 |   17.5781 |        - |     - |   118.77 KB |

```

**Reply by jamescourtney** [reply]
Thanks a lot. I'll give this a shot. These objects are pretty small and cheap, so I can definitely see that Gen0 GC would be quicker than pooling them, especially if the GC is on another thread.

**Comment by jamescourtney** [comment]
I'm beginning to get my head around the idea of parallel reference/value types. Part of me still really thinks that "We're using C# -- allocation and GC are ways of life here". However -- I get that there are Unity users and other folks with true high-performance needs that would benefit from something like this.

However, I don't think FlatSharp is going to be willing/able to do parsing of value-type tables. That's a huge work item, and ends up obscuring a lot of the value that Flatsharp provides (different deserialization modes, etc). Those will need to remain references for the foreseeable future, so please understand that that will probably remain a "won't fix". 

Serializing value-type tables is (relatively) straightforward with some modest refactoring.

Still need to prototype this and think through some things pretty thoroughly (ie, should value tables be allowed to contain reference tables? or is it turtles all the way down?), so please don't take this as a commitment or anything like that.

**Comment by jamescourtney** [comment]
Took a look at your recycling benchmark, and there are some simple ways to improve it. Here are my results:

|            Method |        Mean |       Error |   StdDev | Ratio | RatioSD |     Gen 0 |    Gen 1 | Gen 2 |   Allocated |
|------------------ |------------:|------------:|---------:|------:|--------:|----------:|---------:|------:|------------:|
|         Class1000 | 10,971.2 us | 1,054.35 us | 57.79 us | 11.83 |    0.25 | 2875.0000 | 171.8750 |     - | 47001.43 KB |
| Class1000Recycled |  5,592.2 us |   567.39 us | 31.10 us |  6.03 |    0.15 | 1289.0625 |  93.7500 |     - | 21072.28 KB |
|        Struct1000 |  2,148.3 us |    57.66 us |  3.16 us |  2.32 |    0.06 |         - |        - |     - |    25.15 KB |
|  DataProvider1000 |    927.5 us |   441.60 us | 24.21 us |  1.00 |    0.00 |    5.8594 |        - |     - |   118.77 KB |

Here's the code. There are two sources of inefficiency in the pooling approach. The first is that repeatedly accessing `ThreadLocal.Value` is slow. It's much better to pull the value once and then use it as a local. That alone made pooling as fast as allocating new instances each time (and -- I suspect in a real server under memory pressure would be considerably faster). The second thing is to not clear the list on each pass. Instead, it's quite a bit cheaper to wrap the "finished" list inside an IList<T> wrapper that has a fixed length that's <= the length of the real list.

Based on these numbers, I further suspect that most of the inefficiencies you see now are due to the Union type being a class and not a struct. I have a feeling that if you combined pooling and value unions, you'd see numbers roughly on par with the struct-based approach. 

```c#

        private static readonly ThreadLocal<List<KeyValue>> SimplePool = new(() => new List<KeyValue>());

        [Benchmark]
        public void Class1000Recycled()
        {
            var kvs = SimplePool.Value;
            var buffer = new byte[1024];
            var now = DateTime.Now.ToFileTimeUtc();
            var rowsCount = rows.Length;
            var headerCount = header.Length;

            for (int r = 0; r < 1000; r++)
            {
                int count = 0;
                var row = rows[r % rowsCount];
                for (int col = 0; col < row.Length && col < headerCount; col++)
                {
                    string rowValue = row[col];
                    if (string.IsNullOrWhiteSpace(rowValue))
                    {
                        continue;
                    }

                    KeyValue kv;
                    if (count < kvs.Count)
                    {
                        kv = kvs[count];
                    }
                    else
                    {
                        kv = new();
                        kvs.Add(kv);
                    }

                    kv.Key = new Primitive(header[col]);
                    kv.Value = new Primitive(rowValue);
                    kv.Timestamp = now;
                    count++;
                }

                ListWithKnownLength<KeyValue> list = new ListWithKnownLength<KeyValue>(kvs, count);
                var map = new Map
                {
                    Items = list
                };

                // serialize Kvs
                //var writtenSize = serialize(map, buffer);

                //kvs.Clear();
            }
        }

        private class ListWithKnownLength<T> : IList<T>
        {
            private List<T> inner;
            private int length;

            public ListWithKnownLength(List<T> inner, int length)
            {
                this.inner = inner;
                this.length = length;
            }

            public T this[int index] 
            { 
                get => inner[index];
                set => throw new NotImplementedException(); 
            }

            public int Count => this.length;

            public bool IsReadOnly => true;

            public void Add(T item)
            {
                throw new NotImplementedException();
            }

            public void Clear()
            {
                throw new NotImplementedException();
            }

            public bool Contains(T item)
            {
                return inner.Contains(item);
            }

            public void CopyTo(T[] array, int arrayIndex)
            {
                inner.CopyTo(array, arrayIndex);
            }

            public IEnumerator<T> GetEnumerator()
            {
                return inner.GetEnumerator();
            }

            public int IndexOf(T item)
            {
                return inner.IndexOf(item);
            }

            public void Insert(int index, T item)
            {
                throw new NotImplementedException();
            }

            public bool Remove(T item)
            {
                throw new NotImplementedException();
            }

            public void RemoveAt(int index)
            {
                throw new NotImplementedException();
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                return ((IEnumerable)inner).GetEnumerator();
            }
        }
```

**Reply by Astn** [reply]
Very cool stuff here!

> I have a feeling that if you combined pooling and value unions, you'd see numbers roughly on par with the struct-based approach.

Ya, I was curious what the cost was of the KeyValue vs the Unions it contains.  ValueType Unions could be huge!

**Reply by jamescourtney** [reply]
Numbers with value unions based off my prototype:

|            Method |       Mean |     Error |   StdDev | Ratio | RatioSD |     Gen 0 |    Gen 1 | Gen 2 |  Allocated |
|------------------ |-----------:|----------:|---------:|------:|--------:|----------:|---------:|------:|-----------:|
|         Class1000 | 7,461.4 us | 629.86 us | 34.52 us |  8.55 |    0.07 | 2492.1875 | 132.8125 |     - | 40735.8 KB |
| Class1000Recycled | 4,688.9 us | 505.81 us | 27.73 us |  5.37 |    0.03 |         - |        - |     - |   55.71 KB |
|        Struct1000 | 2,180.2 us | 107.91 us |  5.91 us |  2.50 |    0.01 |         - |        - |     - |   25.15 KB |
|  DataProvider1000 |   873.1 us |  49.51 us |  2.71 us |  1.00 |    0.00 |    6.8359 |        - |     - |  118.77 KB |

Some observations:
- I think the struct benchmark is not a reasonable bar for comparison. It's just adding a `ValueTuple` to a list, not a value type embedded in another class in a list.
- There actually isn't a huge difference between reference and value unions (5.5us vs 4.7us). That's really not a big change at all (especially given the number of operations that are running here). However, the story might be different if the machine was under memory pressure and unions were leaking into Gen2. However, my sense is that if that's your problem, then the unions aren't going to be your biggest worry. As long as you are using them ephemerally, reference unions seems more than fast enough to me.
- The cost of serializing the data is going to dominate the cost of initializing it. Adding a call to serialize per row in the benchmark for Class1000Recycled bumps the runtime to 28,500us and the Class1000 to ~31,000us (interestingly -- the Class1000 makes up some ground here, since FlatSharp devirtualizes vectors, so `List<T>` is much faster than `SomeCustomList<T>`)

So, with serialization added in, the final results are:

|            Method |        Mean |       Error |    StdDev | Ratio | RatioSD |     Gen 0 |    Gen 1 | Gen 2 |   Allocated |
|------------------ |------------:|------------:|----------:|------:|--------:|----------:|---------:|------:|------------:|
|         Class1000_ReferenceUnion | 35,566.1 us |   859.01 us |  47.09 us | 40.45 |    0.16 | 2857.1429 | 142.8571 |     - | 47023.86 KB |
| Class1000Recycled_ReferenceUnion | 30,169.3 us | 4,304.52 us | 235.95 us | 34.31 |    0.10 | 1281.2500 |  93.7500 |     - |  21094.7 KB |
|         Class1000_ValueUnion | 31,952.2 us | 11,121.08 us | 609.58 us | 32.40 |    0.35 | 2437.5000 | 125.0000 |     - | 40758.23 KB |
| Class1000Recycled_ValueUnion | 27,732.9 us |    999.34 us |  54.78 us | 28.13 |    0.26 |         - |        - |     - |    78.13 KB |

Not sure if Value unions are worth the 10-ish% gain or not. It's definitely something to do eventually, but I'm not seeing anything empirical to make me think it's worth the breaking change. They'll be part of FlatSharp 6, which will either be an accumulation of breaking changes or some major new feature. Probably both.

**Reply by Astn** [reply]
Nice work!
Your machine is quite a bit faster then mine. Looks like a good 2-3x. Especially in the allocations department. 
It makes me wonder how sensitive the object allocations are to memory speed.
I was seeing around 30ms for building the data. And around 30ms for serializing the data.
I'm on a laptop with 64gb of 2666mhz in 2 slots. And a 6 core mobile xeon (E-2176M). 
Do you know what you have?

**Reply by jamescourtney** [reply]
> Do you know what you have?

I'm a bit of a PC hardware hobbyist. I have a 5950x (16 core) and 64GB in a 16x4 configuration at 3800mhz.

**Reply by Astn** [reply]
Dang! Nice setup man. Well double the memory speed in double the channels could explain why we see different ratios between Ref and Value types.

**Reply by Astn** [reply]
>     * It's just adding a `ValueTuple` to a list, not a value type embedded in another class in a list.

Too true. I suppose I was thinking about value unions being nearly transparent boxes around things like strings and Ints. Then a Two Tuple being logically coercible into a KeyValue, as they are both just supposed to be simple containers.
So that definitely representing a best case scenario, and something that would require a fair bit of customization in the serializer path.
Is that something your plugin model you've been talking about would be able to do?

**Reply by jamescourtney** [reply]
It could help with that. FlatSharp is at least nominally extensible via `ITypeModel` and `ITypeModelProvider`. `ITypeModel` is the interface that "understands" the flatbuffer schema and generates the serialize/parse/clone/getmaxsize code. However, this is currently only supported with the runtime configuration, not through the compiler. If you wanted to add `ValueTupleTypeModel`, you'd be free to do so. The interface isn't exactly easy to implement, and requires some deep knowledge of FlatBuffers to do so successfully.

I think a plugin system would be neat, but I'm still not certain it's something I *should* build. My thinking on that is that most people using FBS schemas probably want interop with other languages, something that a plugin system works against.
